packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js

```diff
@@ -618,4 +618,27 @@ describe('Shared useSyncExternalStore behavior (shim and built-in)', () => {
       expect(root).toMatchRenderedOutput('A1B1');
     });
   });
+
+  test('Infinite loop if getSnapshot keeps returning new reference', () => {
+    const store = createExternalStore({});
+
+    function App() {
+      const text = useSyncExternalStore(store.subscribe, () => ({}));
+      return <Text text={JSON.stringify(text)} />;
+    }
+
+    spyOnDevAndProd(console, 'error')
+
+    expect(() => {
+      act(() => {
+        createRoot(<App />);
+      })
+    })
+    .toThrow(
+      'Maximum update depth exceeded. This can happen when a component repeatedly ' +
+      'calls setState inside componentWillUpdate or componentDidUpdate. React limits ' + 
+      'the number of nested updates to prevent infinite loops.'
+    )
+    expect(console.error.calls.argsFor(0)[0]).toMatch('The result of getSnapshot should be cached to avoid an infinite loop')
```

Comment:
In CI, we run our tests in production, too, so this will. You can wrap this assertion with `if (__DEV__)`.